proposals/route-blinding.md

```diff
@@ -0,0 +1,285 @@
+is useful when retrying a failed route or using multi-part payments).
+
+The downside compared to rendezvous is that senders have more leeway to probe by changing various
+variables, so the scheme needs to explicitly defend against probing attacks and may be less private.
+
+Some use-cases where route blinding is useful include:
+
+* Recipient anonymity when receiving payments
+* Using unannounced channels in invoices without revealing them
+* Forcing a payment to go through a specific set of intermediaries that can witness the payment
+* Providing anonymous reply paths for onion messages (offers, stuckless payments, etc)
+
+### Overview
+
+At a high level, route blinding works by having the recipient choose an introduction point and a
+route to himself from that introduction point. The recipient then blinds each node and channel
+along that route with ECDH. The recipient includes the blinded route and a secret in the invoice,
+which allows each node in the blinded route to incrementally unblind the payloads.
+
+This scheme requires all the nodes in the blinded route and the sender to activate support for the
+feature. It only becomes effective once a big enough share of the network supports it.
+
+### Notations
+
+* A node `N(i)`'s `node_id` is defined as: `P(i) = k(i) * G`.
+* Blinded `node_id`s are defined as: `B(i) = b(i) * G`.
+* Ephemeral public keys are defined as: `E(i) = e(i) * G`.
+
+### Requirements
+
+A node `N(r)` wants to provide a blinded route `N(r) <- ... <- N(1) <- N(0)` that must be used
+to receive onion messages.
+
+* The channels used along that route may be either announced or unannounced.
+* When used for payments, intermediate nodes in the blinded route MUST NOT learn `payment_secret`.
+* Intermediate nodes in the blinded route MUST NOT learn the `node_id` or `scid` of other
+  intermediate nodes except for their immediate predecessor or successor.
+* Intermediate nodes in the blinded route MUST NOT learn their distance to the recipient `N(r)`.
+* Senders MUST NOT learn the real `node_id`s and `scid`s of the blinded intermediate hops after the
+  introduction point `N(0)`.
+* If `N(r)` creates multiple blinded routes to herself, senders MUST NOT be able to tell that these
+  routes lead to the same recipient (unless this information is leaked by higher layers of the
+  protocol, such as using the same `payment_hash`).
+
+### Encrypted recipient data
+
+Route blinding introduces a new TLV field to the onion `tlv_payload`: the `encrypted_recipient_data`.
+
+This field is used to carry data coming from the final recipient that cannot be modified by the
+sender. For route blinding it only needs to contain the `scid` to use when forwarding the message,
+but it may be extended with additional data in the future. It uses ChaCha20-Poly1305 as AEAD scheme.
+
+1. type: 10 (`encrypted_recipient_data`)
+2. data:
+    * [`...*byte`:`encrypted_recipient_data`]
```

Comment:
It is unclear how multiple encrypted recipient data destined for intermediate hops is serialized. I'd guess they are nested, i.e., `N(i)` decodes its data, which results in a TLV stream, which itself contains the encrypted data for `N(i+1)`, but I could not find it defined anywhere.

The alternative is to have multiple TLV fields with type=10, one for each hop, which presumably would also be ok, but then we need a way for each hop to identify the field it should use (we can't just attach them in order due to the lexicographic ordering requirement for TLV serialization).

Reply:
> I'd guess they are nested, i.e., N(i) decodes its data, which results in a TLV stream, which itself contains the encrypted data for N(i+1), but I could not find it defined anywhere.

I'm not sure that would work, as it means that intermediate nodes would need to modify the sphinx payload before forwarding (IIUC).

With the current proposal, there is one `encrypted_recipient_data` in each Bolt 4 `hop_payload`. `N(i)` finds one such TLV in his `hop_payload`, containing only the data for `N(i)`. The `encrypted_recipient_data` for `N(i+1)` is in the next `hop_payload`, completely separate. Does that make sense? I think the only way to make this more explicit would be to improve my ASCII art below to show the whole onion at each step.

Reply:
Done in https://github.com/lightningnetwork/lightning-rfc/pull/765/commits/0468ed01e003e1e1ed7e49c0802b5f4c10519cba